File: 0347-top-k-frequent-elements/0347-top-k-frequent-elements.cpp
class Solution {
public:
typedef pair<int, int> pi;
    vector<int> topKFrequent(vector<int>& nums, int k) {
        unordered_map<int, int> mp;
        for (auto a : nums) {
            mp[a]++;
        }
        priority_queue<pi, vector<pi>, greater<pi>> pq;

        for (auto [ele, freq] : mp) {
            pair<int,int> p = {freq, ele};
            pq.push(p);
            if (pq.size() > k)
                pq.pop();
        }

        vector<int> ans;
        while(!pq.empty()){
            int ele = pq.top().second;
            ans.push_back(ele);
            pq.pop();
        }

        return ans;
    }
};

File: 0347-top-k-frequent-elements/0347-top-k-frequent-elements.py
class Solution(object):
    def topKFrequent(self, nums, k):
        """
        :type nums: List[int]
        :type k: int
        :rtype: List[int]
        """
        mp = {}
        minHeap = []
        for x in nums:
            if x not in mp:
                mp[x] = 1
            else:
                mp[x] += 1
        
        for item, freq in mp.items():
            heapq.heappush(minHeap, (freq, item))
            if len(minHeap) > k:
                heapq.heappop(minHeap)
        return [item for (freq, item) in minHeap]
        